### ISSUE TITLE
No output from pyplot on cygwin64 python3 virtualenv

### ISSUE THREAD

**Issue by s7726** [matplotlib/matplotlib#5962]
Encountered initially trying to work with:
https://github.com/amueller/word_cloud/blob/master/examples/simple.py

Then tried the matplotlib xkcd example.

I run the file `python xkcd.py` and the command runs, and returns (0 exit status), but no windows are displayed and no errors are given.

I'm running the latest cygwin x64 with a virtualenv running python 3.4.3 and matplotlib 1.5.1

I'm willing to help hunt down the issue. But I don't know where to start.

**Comment by WeatherGod** [comment]
How did you install/build matplotlib? Chances are, if you built it from
source, then you didn't have the gui headers available, so it fell back to
a default of the headless AGG backend.

**Comment by s7726** [comment]
I installed using `pip install matplotlib`. Is there a particular set of libraries or dev packages I should be looking for to install?
Some kind of warning or something would be nice, when it's defaulting to a perceived inoperable mode.

**Comment by tacaswell** [comment]
What does

```python
import matplotlib
print(matplotlib.get_backend())
```

give you?  My _guess_ is that it is going to say 'Agg', which is a non interactive backend.  My guess is that you are missing a gui-toolkit wrapping library.

**Comment by s7726** [comment]
`Qt5Agg`

What should I have available for it to use something else?
It would be nice if there were some kind of prompt or message to indicate defaulting to a non-functional (from user perspective) behavior.

**Comment by tacaswell** [comment]
Are you using plain python or ipython with `%matplotlib` magic.

There are many use cases for mpl where an interactive backend is _not_ desired so we can not guess when the lack of one is intentional or not.

Does pyqt5 import cleanly?

Try

```
import matplotlib
import matplotlib.pyplot as plt
print(matplotlib.get_backend())
plt.plot(range(3))
plt.show()
```

**Comment by AmeyaVS** [comment]
I am also facing similar issue under cygwin.

Initially when I installed matplotlib-2.0.0 in python-2.7 which is working fine.
But when I moved to python-3.4.5 suddenly the sample applications also stopped working.
After looking into dependencies and found out that the python3-tkinter backend was missing.

After installing all the dependencies and rebuilding I am getting this error:
```python3
>>> import matplotlib.pyplot as plt
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "$HOME/.local/lib/python3.4/site-packages/matplotlib/pyplot.py", line 115, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "$HOME/.local/lib/python3.4/site-packages/matplotlib/backends/__init__.py", line 32, in pylab_setup
    globals(),locals(),[backend_name],0)
  File "$HOME/.local/lib/python3.4/site-packages/matplotlib/backends/backend_tkagg.py", line 13, in <module>
    import matplotlib.backends.tkagg as tkagg
  File "$HOME/.local/lib/python3.4/site-packages/matplotlib/backends/tkagg.py", line 9, in <module>
    from matplotlib.backends import _tkagg
RuntimeError: No such process
```

After poking around my local installation for both python-2.7 and python-3.4.5 packages these are the differences I found in the libraries built:

Python-2.7:
```sh
$HOME/.local/lib/python2.7/site-packages/matplotlib/backends/_tkagg.dll
```

Python-3.4.5:
```sh
$HOME/.local/lib/python3.4/site-packages/matplotlib/backends/_tkagg.cpython-34m.dll
```

It seems while importing the _tkagg package it looks for the _tkagg.dll

Is there a way I can suppress the "cpython-34m" in the generated library? or
Should I file a bug request at Cygwin for reporting this behaviour for the build?

**Note:** I am using pip3 for managing the python libraries.

**Comment by KptainBiwouak** [comment]
I am facing exactly same kind of issue after trying to build Matplolib (v2.0.2 & v2.1.0).
Build is successfull but Import _tkagg fails

Could you provide us some guidelines in order to be able to use this module?
Which dependencies are exactly required?
Note: I am working from Cygwin 2.9.0 (x64) + Python 3.6.1

**Comment by matthew-brett** [comment]
@KptainBiwouak - what do you get for:

```
>>> import tkinter
```

in Python?

**Comment by KptainBiwouak** [comment]
Using Python 3.6.1, I got no error.
It works. 
I am using default python3-tkinter-3.6.1-1  module provided from Cygwin packages

**Comment by lharcke** [comment]
~~~~

$ uname -a
CYGWIN_NT-10.0 <hostname> 2.10.0(0.325/5/3) 2018-02-02 15:16 x86_64 Cygwin

$ ipython3
Python 3.6.4 (default, Jan  7 2018, 15:53:53)

In [1]: matplotlib.__version__
Out[1]: '2.2.2'
In [7]: import matplotlib.pyplot as plt
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-7-eff513f636fd> in <module>()
----> 1 import matplotlib.pyplot as plt

/home/lharcke/.local/lib/python3.6/site-packages/matplotlib/pyplot.py in <module>()
    113 ## Global ##
    114
--> 115 _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
    116
    117 _IP_REGISTERED = None

/home/lharcke/.local/lib/python3.6/site-packages/matplotlib/backends/__init__.py in pylab_setup(name)
     60     # imports. 0 means only perform absolute imports.
     61     backend_mod = __import__(backend_name, globals(), locals(),
---> 62                              [backend_name], 0)
     63
     64     # Things we pull in from all backends

/home/lharcke/.local/lib/python3.6/site-packages/matplotlib/backends/backend_tkagg.py in <module>()
      2
      3 from .. import cbook
----> 4 from . import tkagg  # Paint image to Tk photo blitter extension.
      5 from .backend_agg import FigureCanvasAgg
      6 from ._backend_tk import (

/home/lharcke/.local/lib/python3.6/site-packages/matplotlib/backends/tkagg.py in <module>()
      7 import numpy as np
      8
----> 9 from matplotlib.backends import _tkagg
     10
     11 def blit(photoimage, aggimage, bbox=None, colormode=1):

RuntimeError: No such process
~~~~

**Comment by WeatherGod** [comment]
How did you install python and matplotlib?

**Comment by lharcke** [comment]
Python3. Tcl/TK and Tkinter were installed via the Cygwin package manager.  Matplotlib is not packaged for Cygwin, so I did:

`pip3 install --home matplotlib`

**Comment by WeatherGod** [comment]
Hmmm, it is possible that the setup.py did not see the header files for the
cygwin tk libraries, and maybe you ended up compiling against something
different than what it is linking against. This is just speculation.

**Comment by lharcke** [comment]
Remotely possible, as I have an Anaconda 4.2 installation with Tkinter in C:\Program Files\Anaconda3\.  But that would be very weird.  Usually programs running in Cygwin ignore everything outside of the C:\cygwin64\ root directory.  The pseudo-UNIX environment doesn't really permit programs to look above the pseudo-UNIX filesystem root, unless they explicitly try to access the mount point /cygdrive/c/. 
 But UNIX programs generally don't know about the /cygdrive/c/ mount point, unless they are written specifically Cygwin utilities/applications.

Anyway, Matplotlib isn't packaged for Cygwin, and a pip3 install of Matplotlib doesn't work out of the box for some reason (most likely Tk 8.6 incompatibility), which isn't flagged/reported during the pip3 build/install.

**Comment by lharcke** [comment]
So, here's the bit from setupext.py where the TkAgg build is checked:
~~~~
class BackendTkAgg(OptionalBackendPackage):
    name = "tkagg"
    force = True

    def check(self):
        return "installing; run-time loading from Python Tcl / Tk"

    def runtime_check(self):
        """ Checks whether TkAgg runtime dependencies are met
        """
        try:
            import tkinter
        except ImportError:
            return False
        return True

[snip]

    def add_flags(self, ext):
        ext.include_dirs.insert(0, 'src')
        if sys.platform == 'win32':
            # PSAPI library needed for finding Tcl / Tk at run time
            ext.libraries.extend(['psapi'])
        elif sys.platform == 'linux':
            ext.libraries.extend(['dl'])
~~~~
From static inspection, `tkinter` will load when `runtime_check()` is called.  But it would appear that the linker `add_flags()` directive will fail for Cygwin, as on Cygwin, the `sys.platform` returns
~~~~
In [1]: import sys

In [2]: sys.platform
Out[2]: 'cygwin'
~~~~
and `cygwin` is not in the if/then/else check for linker flags.

**Comment by Sonic-Amiga** [comment]
Hello! Just hit the same problem and using the git master branch figured out what's wrong. Apparently Windows doesn't provide for RTLD_LOCAL / RTLD_GLOBAL semantics , so Cygwin's dlsym() against /usr/lib/python3.6/lib-dynload/_tkinter.cpython-36m-x86_64-cygwin.dll fails to locate Tk_FindPhoto symbol, which is in fact declared in /usr/bin/libtk8.6.dll, despite all the dependencies are in place.
 It looks like we'll have to use Windows-native mechanism for symbol lookup for Cygwin.

**Comment by tacaswell** [comment]
@Sonic-Amiga Do you know what those mechanisms are and if so could you open a Pull Request with the changes?

I don't think we currently test cygwin, but I suspect it would be possible on either appveyor or azure?

**Comment by Sonic-Amiga** [comment]
@tacaswell I've sucessfully verified that you can use native Windows API, the same as for "raw" Win32 platform. In src/_tkagg.cpp you have several #ifdef _WIN32. Those ones related to symbol loading should become #if defined(_WIN32) || defined(__CYGWIN__). And add psapi library for cygwin platform. Yes, Cygwin apps are not advised to use WinAPI directly, but you're using it with care here. :) An alternate solution would be to implement POSIX behavior in cygwin1.dll; and i'm not sure how difficult it would be. At least there's no quick solution to that.
 Bear in mind that GetForegroundWindow/SetForegroundWindow should stay _WIN32-only since Cygwin, being an "honest" POSIX, uses X Window API for graphics.

**Comment by Sonic-Amiga** [comment]
@tacaswell Please consider the PR. I've verified it with xkcd.py; it works.

**Comment by QuLogic** [comment]
Fixed by #17415 (I can't test though.)

**Comment by philgoetz** [comment]
I wrote a Python program that worked under cygwin python7 throughout all of May.  I ran it again today and got the exact same error described here.
I can't tell you what backend it's using because:

```
$ python3
Python 3.7.7 (default, Apr 10 2020, 07:59:19)
[GCC 9.3.0] on cygwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import matplotlib
>>> print(matplotlib.get_backend())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/site-packages/matplotlib/__init__.py", line 1302, in get_backend
    return rcParams['backend']
  File "/usr/lib/python3.7/site-packages/matplotlib/__init__.py", line 831, in __getitem__
    from matplotlib import pyplot as plt
  File "/usr/lib/python3.7/site-packages/matplotlib/pyplot.py", line 2282, in <module>
    switch_backend(rcParams["backend"])
  File "/usr/lib/python3.7/site-packages/matplotlib/__init__.py", line 832, in __getitem__
    plt.switch_backend(rcsetup._auto_backend_sentinel)
  File "/usr/lib/python3.7/site-packages/matplotlib/pyplot.py", line 204, in switch_backend
    switch_backend(candidate)
  File "/usr/lib/python3.7/site-packages/matplotlib/pyplot.py", line 221, in switch_backend
    backend_mod = importlib.import_module(backend_name)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/usr/lib/python3.7/site-packages/matplotlib/backends/backend_tkagg.py", line 1, in <module>
    from . import _backend_tk
  File "/usr/lib/python3.7/site-packages/matplotlib/backends/_backend_tk.py", line 22, in <module>
    from . import _tkagg
RuntimeError: No such process

```
I didn't use Python for anything else since the last time I ran my program.
I did, however, run Cygwin's auto-update several times.
I definitely also ran Cygwin's auto-update in May, before running my program successfully.
So some Cygwin update released in May or June is causing this error again.

**Comment by tacaswell** [comment]
@philgoetz Try re-installing Matplotlib (you have have changed Tk under it?).  If that fails can you try `pip install --upgrade --pre matplotlib` to try out  3.3.0rc1 which has the changes that closed this issue.

**Comment by philgoetz** [comment]
I found this comment [on stackoverflow](https://stackoverflow.com/questions/42401446/error-importing-seaborn-in-cygwin
) by me, describing the earlier problem:

`I'm having the exact same problem, with Python 3.7. Already installed tcl-tk-devel via the cygwin setup. Note the import statement specifies that _tkagg must be in (for me) /usr/lib/python3.7/site-packages/matplotlib/backends . And it is there! File _tkagg.cpython-37m-x86_64-cygwin.dll . For some reason the import fails to load it. – Phil Goetz Apr 2 at 20:50`

And I found this solution, which I had written down in my notes, copied from https://github.com/matplotlib/matplotlib/issues/5962

`It would seem there is a known incompatibility between Matplotlib and Tk version 8.6, but the pip3 install doesn't check and warn about Tk 8.6 incompatibility and disable TkAgg if Tk 8.6 is present when attempting to build/install the TkAgg interface. At least nothing appeared on the command line during a pip3 install --home matplotlib install. It is possible if I threw the -v switch, there might have been some sort of warning generated.`

So I reverted cygwin64 package tk-devel from Tk 8.6 to 8.5.11-1 .
(This reversion didn't work for cygwin 32-bit, BTW; it complained that it couldn't download package 'freeglut'.)
But I kept getting the same _tkagg error.

I tried re-installing matplotlib, and in poking around discovered that my cygwin pip3 is, apparently, installing packages in my Anaconda3 install of python3.7, not my cygwin install of python3.7.  Then I remembered that that had been my solution: I had to run my program inside an Anaconda3 window, using Anaconda's python.  So I did that, and my program worked.  I still haven't been able to get it run with Cygwin's python3.7 .
